### ISSUE TITLE
Pin SDKs to the version of pulumi that generated them

### ISSUE THREAD

**Issue by blampe** [pulumi/pulumi#14789]
Generated node SDKs for example take a dependency on `"@pulumi/pulumi": "^3.42.0"`:

https://github.com/pulumi/pulumi/blob/master/pkg/codegen/nodejs/gen.go#L49

But this is fragile -- a new pu/pu release can inadvertently break provider SDK generation. See https://github.com/pulumi/pulumi/issues/14781 for an example.

Instead, the SDK should pin to the version of pulumi that generated it.

If we're generating with a development version of pulumi (for example as part of a downstream test) it would be enormously useful to be able to pin instead to a local pulumi SDK.

In other words:
* pulumi version 3.xy.z -> generate provider SDKs pinned to 3.xy.z
* pulumi dev → generate provider SDKs pointed at `../../pulumi/sdk/langauge/`

Note this enhancement is relevant to to all our languages, not just node (but except maybe YAML? I'm not sure how that works).
